include/autoppl/algorithm/mh.hpp

```diff
@@ -0,0 +1,123 @@
+#pragma once
+#include <chrono>
+#include <random>
+#include <algorithm>
+#include <vector>
+#include <autoppl/util/traits.hpp>
+#include <autoppl/variable.hpp>
+#include <autoppl/expression/distribution/base.hpp>
+
+/*
+ * Assumptions:
+ * - every variable referenced in model is of type Variable<double>
+ */
+
+namespace ppl {
+namespace details {
+
+struct MHData
+{
+    double next;
+    // TODO: maybe keep an array for batch sampling?
+};
+
+} // namespace details
+
+/*
+ * Metropolis-Hastings algorithm to sample from posterior distribution.
+ * The posterior distribution is a constant multiple of model.pdf().
+ * Any variables that model references which are in state "parameter"
+ * is sampled and in state "data" are not.
+ * So, model.pdf() is proportional to p(parameters... | data...).
+ *
+ * User must ensure that they allocated at least as large as n_sample
+ * in the storage associated with every parameter referenced in model.
+ */
+template <class ModelType>
+inline void mh_posterior(ModelType& model,
+                         double n_sample,
+                         double stddev = 1.0,
+                         double seed = std::chrono::duration_cast<
+                                        std::chrono::milliseconds>(
+                                            std::chrono::system_clock::now().time_since_epoch()
+                                            ).count()
+                        )
+{
+    using data_t = details::MHData;
+    
+    // set-up auxiliary tools
+    std::mt19937 gen(seed);
+    std::uniform_real_distribution unif_sampler(0., 1.);
+
+    // get number of parameters to sample    
+    size_t n_params = 0.;
+    auto get_n_params = [&](auto& eq_node) {
+        auto& var = eq_node.get_variable();
+        using var_t = std::decay_t<decltype(var)>;
```

Comment:
Can you explain this idiom? What does std::decay_t<decltype(var)>; do?

Reply:
(reposting for record) yup basically gets rid of references and const/volatile qualifiers to get the "underlying" type. It's needed because when you try to grab member alias of a type T, but T is like A&, you can't do T::value_t even. Has to be like A::value_t.